src/Microsoft.AspNet.OData.Shared/Extensions/ODataPathExtensions.cs

```diff
@@ -11,46 +11,75 @@
 
 namespace Microsoft.AspNet.OData.Extensions
 {
-    internal static class ODataPathExtensions
+    /// <summary>
+    /// Extensions method for <see cref="ODataPath"/>.
+    /// </summary>
+    public static class ODataPathExtensions
     {
+        /// <summary>
+        /// Get keys from the last <see cref="KeySegment"/>.
+        /// </summary>
+        /// <param name="path"><see cref="ODataPath"/>.</param>
+        /// <returns>Dictionary of keys.</returns>
         public static Dictionary<string, object> GetKeys(this ODataPath path)
         {
             if (path == null)
             {
                 throw Error.ArgumentNull(nameof(path));
             }
 
-            Dictionary<string, object> keys = new Dictionary<string, object>();
+            List<ODataPathSegment> pathSegments = path.AsList();
+            int position = path.Count - 1;
+            ODataPathSegment pathSegment = pathSegments[position];
```

Comment:
This could be potentially assign a segment that is not a `KeySegment` to `pathSegment` and then we'll end up calling `GetKeysFromKeySegment` with a null value. That would actually be a problem because `GetKeysFromKeySegment` actually throws an exception if a null is passed.
We should rewrite the logic such that the value assigned to the `pathSegment` variable can only be a `KeySegment` or null.
In addition, we should never call `GetKeysFromKeySegment` if we didn't find a `KeySegment`

Reply:
Fixed and added a test